### PR TITLE
 plan: fix a bug in decorrelation

### PR DESCRIPTION
**Pull request by winoros**
## What have you changed? (mandatory)

Previous, if the decorrelation failed. The equal condition which contains the correlated column is removed.
This pr is to fix it.

## What are the type of the changes (mandatory)?
- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested (mandatory)?

unit-test

### CODE REVIEW

File: plan/rule_decorrelate.go [hunk 1]
```diff
@@ -202,35 +202,38 @@ func (s *decorrelateSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
...
-								agg.schema.Append(clonedCol.(*expression.Column))
-							}
-							// If group by cols don't contain the join key, add it into this.
-							if agg.getGbyColIndex(eqCond.GetArgs()[1].(*expression.Column)) == -1 {
-								agg.GroupByItems = append(agg.GroupByItems, clonedCol)
-							}
+				originalExpr := sel.Conditions
+				sel.Conditions = remainedExpr
+				// There's no other correlated column.
+				if apply.extractCorColumnsBySchema(); len(eqCondWithCorCol) > 0 && len(apply.corCols) == 0 {
```
Comment (zz-jason) [inline plan/rule_decorrelate.go]: better to move `apply.extractCorColumnsBySchema()` outside the `if` statement

File: plan/rule_decorrelate.go [hunk 2]
```diff
@@ -202,35 +202,38 @@ func (s *decorrelateSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
...
-								newFunc := aggregation.NewAggFuncDesc(apply.ctx, ast.AggFuncFirstRow, []expression.Expression{clonedCol}, false)
-								agg.AggFuncs = append(agg.AggFuncs, newFunc)
-								agg.schema.Append(clonedCol.(*expression.Column))
-							}
-							// If group by cols don't contain the join key, add it into this.
-							if agg.getGbyColIndex(eqCond.GetArgs()[1].(*expression.Column)) == -1 {
-								agg.GroupByItems = append(agg.GroupByItems, clonedCol)
-							}
+				originalExpr := sel.Conditions
+				sel.Conditions = remainedExpr
```
Comment (coocood) [inline plan/rule_decorrelate.go]: Can we not modify `sel.Conditions` here?
If we check `len(eqCondWithCorCol) == len(sel.Conditions)`, we can determine if `selection can be removed, and `remainedExpr` is not needed.

Comment (winoros) [reply plan/rule_decorrelate.go]: There can be the case that the plan below this selection has correlated column. It's not normal case. But it still exists.

File: plan/rule_decorrelate.go (hunk 1 above)
Comment (winoros) [reply plan/rule_decorrelate.go]: So move to where?

Comment (coocood) [inline plan/rule_decorrelate.go]: Keep the original intent is easier to review.

File: plan/rule_decorrelate.go [hunk 3]
```diff
@@ -202,14 +202,20 @@ func (s *decorrelateSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
...
-					if expr := apply.deCorColFromEqExpr(sel.Conditions[i]); expr != nil {
+				for _, cond := range sel.Conditions {
+					if expr := apply.deCorColFromEqExpr(cond); expr != nil {
 						eqCondWithCorCol = append(eqCondWithCorCol, expr.(*expression.ScalarFunction))
-						sel.Conditions = append(sel.Conditions[:i], sel.Conditions[i+1:]...)
+					} else {
+						remainedExpr = append(remainedExpr, cond)
 					}
 				}
+				originalExpr := sel.Conditions
```
Comment (coocood) [inline plan/rule_decorrelate.go]: Can we move this two lines in `if len(eqCondWithCorCol) > 0`?

**Comment by coocood** [comment]
LGTM

**Pull request merged**